### PR TITLE
Enabling ROPC on CCA

### PR DESCRIPTION
**Pull request by trwalke**
Fixes #
https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/3774

**Changes proposed in this request**
Adding api to enable ROPC on CCA App

**Testing**
Integration

**Performance impact**

**Documentation**
- [ ] All relevant documentation is updated.

### CODE REVIEW

File: src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenByUsernameAndPasswordConfidentialParameterBuilder.cs [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+using Microsoft.Identity.Client.PlatformsCommon.Shared;
+using Microsoft.Identity.Client.TelemetryCore.Internal.Events;
+
+namespace Microsoft.Identity.Client
+{
+    /// <summary>
+    /// Parameter builder for the <see cref="IPublicClientApplication.AcquireTokenByUsernamePassword(IEnumerable{string}, string, string)"/>
+    /// operation. See https://aka.ms/msal-net-up
+    /// </summary>
+    public sealed class AcquireTokenByUsernameAndPasswordConfidentialParameterBuilder :
```
Comment (trwalke) [inline src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenByUsernameAndPasswordConfidentialParameterBuilder.cs:24]: The issue recommends against the moving of the current api to the base class and creating a new one. Not entirely sure this is the best way unless there is some reason you were thinking of? Perhaps I misunderstood what you mean exactly? Thoughts on this @bgavrilMS

Comment (bgavrilMS) [reply src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenByUsernameAndPasswordConfidentialParameterBuilder.cs:24]: Well we have AcquireTokenSilent on the base class and it's a struggle to reason about it - we keep having to check `IsConfidentialClient` etc.

I'm thinking more for the future PCA / CCA split.

File: src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenByUsernameAndPasswordConfidentialParameterBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+                   .WithScopes(scopes).WithUsername(username).WithPassword(password);
+        }
+
+        /// <summary>
+        /// Enables MSAL to read the federation metadata for a WS-Trust exchange from the provided input instead of acquiring it from an endpoint.
+        /// This is only applicable for managed ADFS accounts. See https://aka.ms/MsalFederationMetadata.
+        /// </summary>
+        /// <param name="federationMetadata">Federation metadata in the form of XML.</param>
+        /// <returns>The builder to chain the .With methods</returns>
+        public AcquireTokenByUsernameAndPasswordConfidentialParameterBuilder WithFederationMetadata(string federationMetadata)
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenByUsernameAndPasswordConfidentialParameterBuilder.cs:50]: Let's please not do SAML flows for CCA.

File: src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenByUsernameAndPasswordConfidentialParameterBuilder.cs [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+        /// </list>
+        /// </remarks>
+#if iOS || ANDROID
+        [EditorBrowsable(EditorBrowsableState.Never)]
+#endif
+        public AcquireTokenByUsernameAndPasswordConfidentialParameterBuilder WithProofOfPossession(string nonce, HttpMethod httpMethod, Uri requestUri)
+        {
+            ClientApplicationBase.GuardMobileFrameworks();
+
+            if (!ServiceBundle.Config.IsBrokerEnabled)
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenByUsernameAndPasswordConfidentialParameterBuilder.cs:82]: See this is where it starts to be strange. We're mixing PCA concepts in here. 

We could support ROPC + confidential client POP in reality, no?

File: src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenByUsernameAndPasswordConfidentialParameterBuilder.cs [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+        internal AcquireTokenByUsernameAndPasswordConfidentialParameterBuilder(IConfidentialClientApplicationExecutor confidentialClientApplicationExecutor)
+            : base(confidentialClientApplicationExecutor)
+        {
+        }
+
+        internal static AcquireTokenByUsernameAndPasswordConfidentialParameterBuilder Create(
+            IConfidentialClientApplicationExecutor confidentialClientApplicationExecutor,
+            IEnumerable<string> scopes,
+            string username,
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenByUsernameAndPasswordConfidentialParameterBuilder.cs:37]: Let's be more agreesive here than on PCA. Do null checks on the params. We're not pulling the username from WIndows if the user doesn't give it to us.

File: src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenByUsernameAndPasswordConfidentialParameterBuilder.cs [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+            popConfig.Nonce = nonce;
+            popConfig.HttpMethod = httpMethod;
+
+            CommonParameters.PopAuthenticationConfiguration = popConfig;
+            CommonParameters.AuthenticationScheme = new PopBrokerAuthenticationScheme();
+
+            return this;
+        }
+
+        private AcquireTokenByUsernameAndPasswordConfidentialParameterBuilder WithUsername(string username)
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenByUsernameAndPasswordConfidentialParameterBuilder.cs:110]: pointless methods :)

File: src/client/Microsoft.Identity.Client/Microsoft.Identity.Client.csproj [hunk 6]
```diff
@@ -86,6 +86,8 @@
     <Compile Include="$(PathToMsalSources)\**\*.cs" Exclude="$(PathToMsalSources)\obj\**\*.*" />
     <Compile Remove="$(PathToMsalSources)\Platforms\**\*.*;$(PathToMsalSources)\Resources\*.cs" />
     <Compile Remove="$(PathToMsalSources)\PlatformsCommon\PlatformNotSupported\ApiConfig\SystemWebViewOptions.cs" />
+    <None Remove="ApiConfig\AcquireTokenByUsernameAndPasswordConfidentialParameterBuilder.cs" />
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/Microsoft.Identity.Client.csproj:89]: undo

File: src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenByUsernameAndPasswordConfidentialParameterBuilder.cs (hunk 3 above)
Comment (trwalke) [reply src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenByUsernameAndPasswordConfidentialParameterBuilder.cs:82]: Yes, this broker code is only for PCA, will remove. I do think it is possible to perform ROPC + POP using the CCA POP implementation.

File: src/client/Microsoft.Identity.Client/Internal/Requests/UsernamePasswordRequest.cs [hunk 7]
```diff
@@ -117,6 +117,11 @@ private async Task<UserAssertion> FetchAssertionFromWsTrustAsync()
 
             if (userRealmResponse.IsFederated)
```
[the remark this answers is not in the document]
Comment (rayluo) [reply src/client/Microsoft.Identity.Client/Internal/Requests/UsernamePasswordRequest.cs:118]: FYI: I have no strong opinion either way. On the one hand, the "ROPC for CCA" topic never mentioned this level of details about federated accounts, so I simply "copied" all the PCA's ROPC behaviors (including all its quirks such as UserRealm) into CCA. On the other hand, the ROPC's federation behavior was questionable and at one point there was an attempt to refactor or remove it, but that attempt was never funded before; we might as well remove it in this PR.

File: src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenByUsernameAndPasswordConfidentialParameterBuilder.cs (hunk 3 above)
Comment (trwalke) [reply src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenByUsernameAndPasswordConfidentialParameterBuilder.cs]: can confirm the ROPC + POP using the CCA POP flow works

Comment (bgavrilMS) [reply src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenByUsernameAndPasswordConfidentialParameterBuilder.cs]: I expected as much, POP is protocol agnostic.

File: src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenByUsernameAndPasswordConfidentialParameterBuilder.cs [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+            Parameters.Password = password;
+        }
+
+        internal static AcquireTokenByUsernameAndPasswordConfidentialParameterBuilder Create(
+            IConfidentialClientApplicationExecutor confidentialClientApplicationExecutor,
+            IEnumerable<string> scopes,
+            string username,
+            string password)
+        {
+            if (username == null)
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenByUsernameAndPasswordConfidentialParameterBuilder.cs:45]: string.nullorempty

File: src/client/Microsoft.Identity.Client/IConfidentialClientApplication.cs [hunk 9]
```diff
@@ -86,6 +87,21 @@ public partial interface IConfidentialClientApplication : IClientApplicationBase
...
+        /// </summary>
+        /// <param name="scopes">Scopes requested to access a protected API.</param>
+        /// <param name="username">Identifier of the user application requests token on behalf.
+        /// Generally in UserPrincipalName (UPN) format, e.g. <c>john.doe@contoso.com</c></param>
+        /// <param name="password">User password as a string.</param>
+        /// <returns>A builder enabling you to add optional parameters before executing the token request.</returns>
+        /// <remarks>
+        /// Available only for .NET Framework and .NET Core applications. See <see href="https://aka.ms/msal-net-up">our documentation</see> for details.
+        /// </remarks>
+        public AcquireTokenByUsernameAndPasswordConfidentialParameterBuilder AcquireTokenByUsernamePassword(IEnumerable<string> scopes, string username, string password);
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/IConfidentialClientApplication.cs:103]: This is breaking change. You'll have to add it in a similar way you added `AcquireTokenByRefreshToken`

File: src/client/Microsoft.Identity.Client/Microsoft.Identity.Client.csproj [hunk 10]
```diff
@@ -86,6 +86,7 @@
     <Compile Include="$(PathToMsalSources)\**\*.cs" Exclude="$(PathToMsalSources)\obj\**\*.*" />
     <Compile Remove="$(PathToMsalSources)\Platforms\**\*.*;$(PathToMsalSources)\Resources\*.cs" />
     <Compile Remove="$(PathToMsalSources)\PlatformsCommon\PlatformNotSupported\ApiConfig\SystemWebViewOptions.cs" />
+    <Compile Include="ApiConfig\AcquireTokenByUsernameAndPasswordConfidentialParameterBuilder.cs" />
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/Microsoft.Identity.Client.csproj:89]: remove

File: src/client/Microsoft.Identity.Client/Internal/Requests/UsernamePasswordRequest.cs [hunk 11]
```diff
@@ -111,6 +111,12 @@ private async Task<UserAssertion> FetchAssertionFromWsTrustAsync()
                 return null;
             }
 
+            //WsTrust not supported on ROPC
+            if (AuthenticationRequestParameters.AppConfig.IsConfidentialClient)
+            {
+                return null;
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/Internal/Requests/UsernamePasswordRequest.cs:117]: Please add some log statements here.

**Review by bgavrilMS — APPROVED** [review]
LGTM. I would like to understand if **federated** user ROPC over CCA respects the secret or not though. If it doesn't, MSAL should disable the flow.

**Review by bgavrilMS — APPROVED** [review]

**Review by bgavrilMS — APPROVED** [review]

File: src/client/Microsoft.Identity.Client/Internal/Requests/UsernamePasswordRequest.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (bgavrilMS) [reply src/client/Microsoft.Identity.Client/Internal/Requests/UsernamePasswordRequest.cs]: I believe @trwalke excludes ROPC for federated users, so this comment can be closed.

File: src/client/Microsoft.Identity.Client/IByUsernameAndPassword.cs [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+    /// </summary>
+    public interface IByUsernameAndPassword
+    {
+        /// <summary>
+        /// Acquires a token without user interaction using username and password authentication.
+        /// This method does not look in the token cache, but stores the result in it. Before calling this method, use other methods 
+        /// such as <see cref="IClientApplicationBase.AcquireTokenSilent(IEnumerable{string}, IAccount)"/> to check the token cache.
+        /// </summary>
+        /// <param name="scopes">Scopes requested to access a protected API.</param>
+        /// <param name="username">Identifier of the user application requests token on behalf.
```
Comment (neha-bhargava) [inline src/client/Microsoft.Identity.Client/IByUsernameAndPassword.cs:24]:
```suggestion
        /// <param name="username">Identifier of the user, application requests token on behalf of.
```

File: src/client/Microsoft.Identity.Client/IByUsernameAndPassword.cs [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+        /// Generally in UserPrincipalName (UPN) format, e.g. <c>john.doe@contoso.com</c></param>
+        /// <param name="password">User password as a string.</param>
+        /// <returns>A builder enabling you to add optional parameters before executing the token request.</returns>
+        /// <remarks>
+        /// Available only for .NET Framework and .NET Core applications. See <see href="https://aka.ms/msal-net-up">our documentation</see> for details.
```
Comment (neha-bhargava) [inline src/client/Microsoft.Identity.Client/IByUsernameAndPassword.cs:29]: Does the documentation on this link valid for confidential client too?

File: src/client/Microsoft.Identity.Client/Microsoft.Identity.Client.csproj [hunk 14]
```diff
@@ -87,6 +87,7 @@
     <Compile Include="$(PathToMsalSources)\**\*.cs" Exclude="$(PathToMsalSources)\obj\**\*.*" />
     <Compile Remove="$(PathToMsalSources)\Platforms\**\*.*;$(PathToMsalSources)\Resources\*.cs" />
     <Compile Remove="$(PathToMsalSources)\PlatformsCommon\PlatformNotSupported\ApiConfig\SystemWebViewOptions.cs" />
+    <Compile Include="IByUsernameAndPassword.cs" />
```
Comment (neha-bhargava) [inline src/client/Microsoft.Identity.Client/Microsoft.Identity.Client.csproj:90]: Remove?

File: tests/Microsoft.Identity.Test.Integration.netcore/HeadlessTests/PoPTests.NetFwk.cs [hunk 15]
```diff
@@ -274,6 +275,43 @@ public async Task PopTestWithRSAAsync()
...
+            var telemetryClient = new TestTelemetryClient(TestConstants.ClientId);
+            var settings = ConfidentialAppSettings.GetSettings(Cloud.Public);
+            var labResponse = await LabUserHelper.GetDefaultUserAsync().ConfigureAwait(false);
+
+            var confidentialApp = ConfidentialClientApplicationBuilder
+                .Create(settings.ClientId)
+                .WithExperimentalFeatures()
+                .WithAuthority(settings.Authority)
+                .WithClientSecret(settings.GetSecret())
+                .WithTelemetryClient(telemetryClient)
```
Comment (neha-bhargava) [inline tests/Microsoft.Identity.Test.Integration.netcore/HeadlessTests/PoPTests.NetFwk.cs:290]: This was marked obsolete recently. Please exclude using it. Can this be validated using the telemetry header for server side telemetry?

**Review by neha-bhargava — APPROVED** [review]
Approved with comments

File: src/client/Microsoft.Identity.Client/IByUsernameAndPassword.cs (hunk 13 above)
Comment (trwalke) [reply src/client/Microsoft.Identity.Client/IByUsernameAndPassword.cs:29]: I think I can update the doc to mention that CCA is also an option

**Pull request merged**